### ISSUE TITLE
kvs.c:781: setroot_event_send: Assertion `root_dir != NULL' failed.

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#3549]
This assertion failure occurred while running the broker under valgrind's `massif` tool, then running `throughput.py -sxn 4096`

```
flux-broker: kvs.c:781: setroot_event_send: Assertion `root_dir != NULL' failed.
==5421== 
==5421== Process terminating with default action of signal 6 (SIGABRT)
==5421==    at 0x6075FB7: raise (raise.c:51)
==5421==    by 0x6077920: abort (abort.c:79)
==5421==    by 0x6067489: __assert_fail_base (assert.c:92)
==5421==    by 0x6067501: __assert_fail (assert.c:101)
==5421==    by 0xAD14594: kvstxn_apply (kvs.c:781)
==5421==    by 0xAD190BD: wait_runqueue (waitqueue.c:167)
==5421==    by 0xAD1826B: cache_entry_set_dirty (cache.c:110)
==5421==    by 0xAD133DE: content_store_completion (kvs.c:629)
==5421==    by 0x4E7AA72: ev_invoke_pending (ev.c:3372)
==5421==    by 0x4E7EA18: ev_run (ev.c:3775)
==5421==    by 0x4E51682: flux_reactor_run (reactor.c:126)
==5421==    by 0xAD17B27: mod_main (kvs.c:2987)
==5421==    by 0x114B35: module_thread (module.c:211)
==5421==    by 0x59CA6DA: start_thread (pthread_create.c:463)
==5421==    by 0x615871E: clone (clone.S:95)
==5421== 
```

**Comment by grondo** [comment]
Also reported by @dongahn in slack:
> One of my Fluxion performance study crashes at 64 nodes with:
> ```
> flux-broker: kvs.c:781: setroot_event_send: Assertion `root_dir != ((void *)0)' failed.
> 216577081344srun: error: quartz74: task 0: Aborted (core dumped)
>```

**Comment by chu11** [comment]
was able to generate a core dump by simply tweaking the `t5000-valgrind` test to do `./throughput.py -sxn 1024`.  

line number points to

```
        assert (root_dir != NULL); // root entry is always in cache on rank 0
```

which i guess the comment is not true :-(

confirmed this is happening on the "primary" kvs dir and not some alternate namespace.

**Comment by chu11** [comment]
I haven't been able to figure out a definitive cause other than "there's no cache entry" for the root entry, leading to the above assert.  But my suspicion is lots of KVS traffic is occurring in between successive heartbeats, and a cache entry is eventually timed out.

I think the fix will be similar to https://github.com/flux-framework/flux-core/pull/2525, where I need to incref/decref a reference to the cache entry.  Although I'm still not sure where to do this.

**Comment by garlick** [comment]
Including the root directory in the setroot event was an optimization and isn't required for correctness as noted here in `setroot_event_process()`:
```c
    /* Optimization: prime local cache with directory object, if provided
     * in event message.  Ignore failure here - object will be fetched on
     * demand from content cache if not in local cache.
     */
    if (!json_is_null (rootdir))
        prime_cache_with_rootdir (ctx, rootdir);
```

Maybe the code in `setroot_event_send()` could include the directory if it's in cache and set it to JSON null if it isn't rather than asserting?

We could drop the `event_includes_rootdir` global.  That was left over from early experimentation.

One theory about why this may have just started happening:  the `sync_cb()` that expires cache entries now has a timer enforcing a 30s deadline, whereas before it was driven by heartbeat messages only.  If there was a backlog of messages, the heartbeats might have been delayed enough to avoid expiring this entry before it is looked up.

**Comment by garlick** [comment]
Also, I wonder if this value is a bit aggressive. 
```c
/* Expire cache_entry after 'max_lastuse_age' seconds.
 */
const double max_lastuse_age = 10.;
```

**Comment by chu11** [comment]
> One theory about why this may have just started happening: the sync_cb() that expires cache entries now has a timer enforcing a 30s deadline, whereas before it was driven by heartbeat messages only. If there was a backlog of messages, the heartbeats might have been delayed enough to avoid expiring this entry before it is looked up.

I thought about this too, but trying the reproducer before the `sync_cb()` work failed in the same way.

**Comment by chu11** [comment]
I appear to have a fix that works.  If someone is desperate to try it out before I come up for air again, you'll find it here: https://github.com/chu11/flux-core/tree/issue3549_kvs_assert

Given how the API to the `kvstxn` internal library works, I think this change is "right".  I think several of the solutions @garlick has listed above would also work, but it perhaps hides the underlying issue a bit.  But what @garlick mentions above might be worthwhile of cleanup.

> We could drop the event_includes_rootdir global. That was left over from early experimentation.

Instead of a global, are you thinking we just send the `root_dir` all of the time?

> const double max_lastuse_age = 10.;

yeah, 10 seconds does seem sort of aggressive.  30 seconds?  60 seconds?  perhaps a heavy throughput test could be used to see and compare against memory usage.

**Comment by chu11** [comment]
got a branch w/ the fix applied to 0.19.0.  I ran the same reproducer (w/ some tweaks to `throughput.py` since the python bindings are much older).  Verified this issue existed on 0.19.0 and this fix appears to make it disappear.   While I was able to trigger the issue on 0.19.0 pretty easily, I haven't gotten the reproducer to "pass" yet (0.19.0 is way slower than 0.24.0, just realized test basically times out after ~1hr, but there are no errors within the 1 hr) .  

I also did some tests by hand to make sure I didn't mess up and have cache entries leak when looking at `flux module stats kvs`. 

https://github.com/chu11/flux-core/tree/issue3549_v0190_kvs_assert

(P.S. there's a test commit in that branch, don't merge that.  Just cherry-pick the head commit).
